platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/strict/StrictVersionConstraints.java

```diff
@@ -105,7 +106,19 @@ public StrictVersionConstraints intersect(StrictVersionConstraints other) {
         if (other == EMPTY) {
             return EMPTY;
         }
-        return of(ImmutableSet.copyOf(Sets.intersection(modules, other.modules)));
+
+        Set<ModuleIdentifier> smaller = (modules.size() < other.modules.size()) ? modules : other.modules;
```

Comment:
🤔 This might be good to put in its own method, with a name that indicates why it exists.